### PR TITLE
feat: [FEAT-007] 개별 메뉴로 주문 등록

### PR DESCRIPTION
**Pull request by HanSeulChung**
주문 후 Menus Entity stock 수량 감소, stock이 0일 경우 품절으로 셋팅

### 변경사항

**AS-IS**

**TO-BE**
1. 개별 메뉴 주문 기능

### 테스트
 
- [ ] 테스트 코드
- [X] API 테스트

### CODE REVIEW

File: src/main/java/com/chs/cafeapp/cart/controller/CartController.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ */
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/cart")
+public class CartController {
+  private final CartService cartService;
+  private final CartMenuService cartMenuService;
+
+  /**
+   * 장바구니 추가
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/cart/controller/CartController.java:35]: 메서드 위의 주석에는 메서드 시그니처의 Param의 의미와 반환 타입의 정보가 기입되는것이 좋습니다.

File: src/main/java/com/chs/cafeapp/cart/dto/CartMenuDto.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public class CartMenuDto {
+  private long id;
+  private long cartId;
+  private long menuId;
+  private String menuName;
+  private int quantity;
+
+
+  public static List<CartMenuDto> of(List<CartMenu> cartMenuList) {
+    List<CartMenuDto> cartMenuDtoList = new ArrayList<>();
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/cart/dto/CartMenuDto.java:22]: cartMenuList 객체 null check 이후에 객체생성이 좋을 듯 해요.

File: src/main/java/com/chs/cafeapp/cart/dto/CartMenuDto.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    if (cartMenuList != null) {
+      for (CartMenu x : cartMenuList) {
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/cart/dto/CartMenuDto.java:24]: 변수명은 정확이 기재해주세요.

File: src/main/java/com/chs/cafeapp/cart/service/impl/CartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  @Override
+  public CartMenuDto addCart(CartInput cartInput, String userId) {
+
+    User user = userRepository.findByLoginId(userId)
+        .orElseThrow(() -> new RuntimeException("해당 사용자가 없습니다."));
+    Menus menus = menuRepository.findById(cartInput.getMenuId())
+        .orElseThrow(() -> new RuntimeException("해당 메뉴가 존재하지 않습니다."));
+
+    Cart cart = user.getCart();
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/cart/service/impl/CartServiceImpl.java:36]: Optional. ofNullable()을 이용하면 아래 구문을 더 깨끗하게 처리 할 수 있지 않을까요?

```java
  if (cart == null) {
      cart = new Cart();
      cart.setUser(user);
      user.setCart(cart);
    }

```

File: src/main/java/com/chs/cafeapp/order/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  private final OrderRepository orderRepository;
+  private final CartMenusRepository cartMenusRepository;
+  private final OrderedMenuRepository orderedMenuRepository;
+
+  private final CartMenuService cartMenuService;
+
+  @Override
+  public OrderDto orderIndividualMenu(OrderInput orderInput, String userId) {
+    User user = userRepository.findByLoginId(userId)
+        .orElseThrow(() -> new RuntimeException("해당 사용자가 존재하지 않습니다."));
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/order/service/impl/OrderServiceImpl.java:41]: RuntimeException 보단 더 구체적인 예외로 정의해보세요.
- https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html

>[AnnotationTypeMismatchException](https://docs.oracle.com/javase/8/docs/api/java/lang/annotation/AnnotationTypeMismatchException.html), [ArithmeticException](https://docs.oracle.com/javase/8/docs/api/java/lang/ArithmeticException.html), [ArrayStoreException](https://docs.oracle.com/javase/8/docs/api/java/lang/ArrayStoreException.html), [BufferOverflowException](https://docs.oracle.com/javase/8/docs/api/java/nio/BufferOverflowException.html), [BufferUnderflowException](https://docs.oracle.com/javase/8/docs/api/java/nio/BufferUnderflowException.html), [CannotRedoException](https://docs.oracle.com/javase/8/docs/api/javax/swing/undo/CannotRedoException.html), [CannotUndoException](https://docs.oracle.com/javase/8/docs/api/javax/swing/undo/CannotUndoException.html), [ClassCastException](https://docs.oracle.com/javase/8/docs/api/java/lang/ClassCastException.html), [CMMException](https://docs.oracle.com/javase/8/docs/api/java/awt/color/CMMException.html), [CompletionException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionException.html), [ConcurrentModificationException](https://docs.oracle.com/javase/8/docs/api/java/util/ConcurrentModificationException.html), [DataBindingException](https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/DataBindingException.html), [DateTimeException](https://docs.oracle.com/javase/8/docs/api/java/time/DateTimeException.html), [DOMException](https://docs.oracle.com/javase/8/docs/api/org/w3c/dom/DOMException.html), [EmptyStackException](https://docs.oracle.com/javase/8/docs/api/java/util/EmptyStackException.html), [EnumConstantNotPresentException](https://docs.oracle.com/javase/8/docs/api/java/lang/EnumConstantNotPresentException.html), [EventException](https://docs.oracle.com/javase/8/docs/api/org/w3c/dom/events/EventException.html), [FileSystemAlreadyExistsException](https://docs.oracle.com/javase/8/docs/api/java/nio/file/FileSystemAlreadyExistsException.html), [FileSystemNotFoundException](https://docs.oracle.com/javase/8/docs/api/java/nio/file/FileSystemNotFoundException.html), [IllegalArgumentException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalArgumentException.html), [IllegalMonitorStateException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalMonitorStateException.html), [IllegalPathStateException](https://docs.oracle.com/javase/8/docs/api/java/awt/geom/IllegalPathStateException.html), [IllegalStateException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalStateException.html), [IllformedLocaleException](https://docs.oracle.com/javase/8/docs/api/java/util/IllformedLocaleException.html), [ImagingOpException](https://docs.oracle.com/javase/8/docs/api/java/awt/image/ImagingOpException.html), [IncompleteAnnotationException](https://docs.oracle.com/javase/8/docs/api/java/lang/annotation/IncompleteAnnotationException.html), [IndexOutOfBoundsException](https://docs.oracle.com/javase/8/docs/api/java/lang/IndexOutOfBoundsException.html), [JMRuntimeException](https://docs.oracle.com/javase/8/docs/api/javax/management/JMRuntimeException.html), [LSException](https://docs.oracle.com/javase/8/docs/api/org/w3c/dom/ls/LSException.html), [MalformedParameterizedTypeException](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/MalformedParameterizedTypeException.html), [MalformedParametersException](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/MalformedParametersException.html), [MirroredTypesException](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/type/MirroredTypesException.html), [MissingResourceException](https://docs.oracle.com/javase/8/docs/api/java/util/MissingResourceException.html), [NegativeArraySizeException](https://docs.oracle.com/javase/8/docs/api/java/lang/NegativeArraySizeException.html), [NoSuchElementException](https://docs.oracle.com/javase/8/docs/api/java/util/NoSuchElementException.html), [NoSuchMechanismException](https://docs.oracle.com/javase/8/docs/api/javax/xml/crypto/NoSuchMechanismException.html), [NullPointerException](https://docs.oracle.com/javase/8/docs/api/java/lang/NullPointerException.html), [ProfileDataException](https://docs.oracle.com/javase/8/docs/api/java/awt/color/ProfileDataException.html), [ProviderException](https://docs.oracle.com/javase/8/docs/api/java/security/ProviderException.html), [ProviderNotFoundException](https://docs.oracle.com/javase/8/docs/api/java/nio/file/ProviderNotFoundException.html), [RasterFormatException](https://docs.oracle.com/javase/8/docs/api/java/awt/image/RasterFormatException.html), [RejectedExecutionException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/RejectedExecutionException.html), [SecurityException](https://docs.oracle.com/javase/8/docs/api/java/lang/SecurityException.html), [SystemException](https://docs.oracle.com/javase/8/docs/api/org/omg/CORBA/SystemException.html), [TypeConstraintException](https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/TypeConstraintException.html), [TypeNotPresentException](https://docs.oracle.com/javase/8/docs/api/java/lang/TypeNotPresentException.html), [UncheckedIOException](https://docs.oracle.com/javase/8/docs/api/java/io/UncheckedIOException.html), [UndeclaredThrowableException](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/UndeclaredThrowableException.html), [UnknownEntityException](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/UnknownEntityException.html), [UnmodifiableSetException](https://docs.oracle.com/javase/8/docs/api/javax/print/attribute/UnmodifiableSetException.html), [UnsupportedOperationException](https://docs.oracle.com/javase/8/docs/api/java/lang/UnsupportedOperationException.html), [WebServiceException](https://docs.oracle.com/javase/8/docs/api/javax/xml/ws/WebServiceException.html), [WrongMethodTypeException](https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/WrongMethodTypeException.html)

File: src/main/java/com/chs/cafeapp/cart/service/impl/CartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    Cart cart = cartRepository.findById(user.getCart().getId()).orElse(null);
+
+
+    if (cart != null) {
+      List<CartMenuDto> cartMenuDtoList = CartMenuDto.of(cart.getCartMenu());
+      return cartMenuDtoList;
+    }
+
+    return null;
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/cart/service/impl/CartServiceImpl.java:96]: null이 좋은지 new ArrayList<>(); 가 좋을지 고민해보세요.

File: src/main/java/com/chs/cafeapp/cart/dto/CartInput.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package com.chs.cafeapp.cart.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/cart/dto/CartInput.java:7]: 전부 필요한지 한번 확인해보세요.

https://projectlombok.org/features/Data

File: src/main/java/com/chs/cafeapp/order/dto/OrderedMenuDto.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+  private long orderId;
+  private long menuId;
+
+  private int quantity;
+  private int totalPrice;
+
+
+  public static List<OrderedMenuDto> of(List<OrderedMenu> orderedMenus) {
+    List<OrderedMenuDto> orderedMenuDtos = new ArrayList<>();
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/order/dto/OrderedMenuDto.java:26]: orderedMenus null check 이후에 인스턴스 생성이 효율적이겠네요.

File: src/main/java/com/chs/cafeapp/order/entity/Order.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      this.totalQuantity += orderedMenu.getQuantity();
+    }
+  }
+
+  public void setTotalPrice(List<OrderedMenu> orderedMenus) {
+    for (OrderedMenu orderedMenu : orderedMenus) {
+      this.totalPrice += orderedMenu.getMenus().getPrice() * orderedMenu.getQuantity();
+    }
+  }
+  public static Order fromOrderInput(OrderInput orderInput) {
```
Comment (jaymores) [inline src/main/java/com/chs/cafeapp/order/entity/Order.java:77]: Order가 의존하는 객체가 많은데요. Order의 책임을 고민해보세요.

DB로 부터 객체로 '매핑된 Order' 를 가지고 응용을 하는 것이라서, 이를 사용하는 모든 객체의 변환로직이 Order에 포함이 되게 된다면   SRP, OCP 원칙을 어기게 되어요.

실무적으로는 코드 충돌이 많이 발생하게 되죠.

고민해보세요. (orderDto, orderInput 등 포함)

**Review by jaymores — COMMENTED** [review]
수고하셨네요! 👍

File: src/main/java/com/chs/cafeapp/order/entity/Order.java (hunk 9 above)
Comment (HanSeulChung) [reply src/main/java/com/chs/cafeapp/order/entity/Order.java:77]: 헉.. 넵 다시 생각해보겠습니다.!

**Comment by HanSeulChung** [comment]
@jaymores 감사합니다!! 제가 놓친게 많았는데 알려주셔서 감사합니다. 다시 참고해서 수정하겠습니다.

**Comment by HanSeulChung** [comment]
우선 간단히 변경할 수 있는것 우선적으로 refactoring 했습니다. 감사합니다.

**Pull request merged**